Add PV.Name into names of generated GCE/AWS/OpenStack volumes.

jsafrane opened the pull request:
Volume names have now format `<cluster-name>-dynamic-<pv-name>`.

`pv-name` is guaranteed to be unique in Kubernetes cluster, adding
`<cluster-name>` ensures we don't conflict with any running cluster
in the cloud project (`kube-controller-manager --cluster-name=XXX`).

'kubernetes' is the default cluster name.

This requires:
- users that share the same cloud project (typically developers of single
  company) should set unique `cluster-name` or they may get a conflict in a very rare
  case.
- cloud admins should delete stray volumes before starting Kubernetes, otherwise
  Kubernetes may generate one that was left over by previous Kubernetes instance.

I'll be happy to make the PV names more robust if required, the above approach looks robust enough to me.

Just to note, before this patch, volume names were `kube-dynamic-<UUID>` or just `<UUID>`, where the UUID was freshly generated and was not related to any Kubernetes object.

cc: @kubernetes/rh-storage @saad-ali @thockin

jsafrane commented:
Oh, and whole purpose of this exercise is to get friendly names in AWS/GCE/OpenStack GUI instead of UUIDs that say nothing.

saad-ali commented on pkg/volume/util.go:
```diff
@@ -148,3 +148,20 @@ func CalculateTimeoutForVolume(minimumTimeout, timeoutIncrement int, pv *api.Per
+	// cut the "<clusterName>-dynamic" to fit full pvName into maxLength
+	// +1 for the '-' dash
+	if pvLen+1+len(prefix) > maxLength {
+		prefix = prefix[:maxLength-pvLen-1]
```
What if `pvLen + 1 > maxLength`?

jsafrane replied:
Good catch. Currently, it cannot happen - GCE has limit of 63 characters and pv.Name there has 12 characters. All the other clouds have 255 characters limit.

Of course, it's not robust, but cutting `pvName` (guaranteed to be unique) may generate non-unique volume names. And error is not an option. Suggestions?

saad-ali replied:
Non-ideal, but it should be ok, because it shouldn't happen.

saad-ali replied:
nit: Maybe add a comment explaining this, in case someone encounters it.

jsafrane replied:
Comment added, I hope it's not too confusing.

saad-ali commented on pkg/volume/util.go:
```diff
@@ -148,3 +148,20 @@ func CalculateTimeoutForVolume(minimumTimeout, timeoutIncrement int, pv *api.Per
 	return (volumeSizeBytes + allocationUnitBytes - 1) / allocationUnitBytes
 }
+
+// GenerateVolumeName returns a PV name with a random suffix.
```
Comment is not quite accurate. Not a `random suffix`. Suffix is passed in as a parameter.

jsafrane replied:
Sorry, that's a leftover from old version. Fixed.

saad-ali commented on pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go:
```diff
@@ -345,6 +347,11 @@ func newProvisioner(plugin volume.ProvisionableVolumePlugin, claim *api.Persiste
+	}
+
+	if pv != nil {
+		volumeOptions.PVName = pv.Name
```
Where is `pv.Name` set?

jsafrane replied:
In Provisioners of individual volume plugins, currently as `PV.GenerateName = "pv-cinder-"`. (pv-aws-, pv-gce-)

E.g. https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/cinder/cinder.go#L434

saad-ali replied:
> `pv-name` is guaranteed to be unique in Kubernetes cluster

I don't understand how the name is guaranteed to be unique then. For example, won't this make the volume name `someCluster-dynamic-pv-cinder-`?

markturansky replied:
A comment would help.

It is guaranteed unique because we first persist the PV with, for example, `GenerateName: "pv-aws-", objectMeta.GenerateName`

We then want to use that PV name as the resource name to tie them together.

saad-ali replied:
Got it, thanks Mark! Yes, a comment would help.

So one more question: why set the final name in such a roundabout fashion? i.e. `NewPersistentVolumeTemplate()` will set it to `pv-aws-UID` then `CreateVolume(...)` will set it to `clusterName-dynamic-pv-aws-UID`. Why not just set it to the final name directly in `NewPersistentVolumeTemplate()`?

markturansky replied:
What kind of name is guaranteed to be unique in the cluster besides a new UUID?  objMeta.GenerateName allows a human-readable prefix with a suffix that makes a unique name.

Also, once the PV is persisted, we get more than a name.  We get a marker (with a future Provisioning Condition) that makes the provisioning process async.  We use annotations today but that is better expressed as a Condition (can put meaningful status on the CLI).

saad-ali replied:
Makes sense, so we want/expect the name of the PV in the provider to be different then (contain more info then) the name of the API PV object.

jsafrane replied:
I added a note describing where pv.Name comes from.

markturansky replied:
@saad-ali strictly speaking, we are assigning PV.Name from Kube to GCEPD.Name in the provider.  The name field exists on both, so we think it would be beneficial for them to match.

jsafrane commented:
@k8s-bot test this issue: #18281

saad-ali commented:
LGTM

jsafrane committed:
Add PV.Name into names of generated GCE/AWS/OSP volumes.

Volume names have now format <cluster-name>-dynamic-<pv-name>.

pv-name is guaranteed to be unique in Kubernetes cluster, adding
<cluster-name> ensures we don't conflict with any running cluster
in the cloud project (kube-controller-manager --cluster-name=XXX).

'kubernetes' is the default cluster name.

jsafrane commented:
Rebased

jsafrane commented:
@k8s-merge-robot, I hate you!

markturansky commented:
Re-adding LGTM after rebase. 'twas already labelled thusly.
